### DISCUSSION TITLE
[proposal] statically implement dynamic components

### DISCUSSION THREAD

**Discussion by DRiFTy17** [tyler-technologies-oss/forge#73] [Ideas]
# Proposal

Refactor all dynamic components to allow for using a "fixed" positioning strategy by default to allow for using statically inline
without needing to dynamically create the element instance and append to the DOM (defaults to the <body> currently).

This would allow for content to stay inline in the location of the DOM where it was added, but render relative to the <body>.
The biggest benefit for this change is to ensure that the content of these floating elements stays intact in its original location
making this much easier to use with frameworks.

We could technically support both the existing usage as well as the new static usage, with static being the default as it's more intuitive and easier to manage.

## Affected components:

- Dialog
- Popup
- Drawer (partially)
- Toast
- Tooltip
- Bottom sheet

When it comes to components like the dialog or bottom-sheet, this is already feasible to support currently, but we should
adjust our default expectation to use these components statically rather than dynamically.

## Benefits

- MUCH easier to provide custom content inline in the HTML regardless of framework being used as the content doesn't need to be moved or re-rendered when these elements are displayed.
- No more requirement of things like `DialogService` in Angular, or custom `<ForgeDialog>` wrapper element in React, but we could still support those for people that want to use it in true dynamic fashion.

## Caveats

One potential concern is that the static usage enables bindings to potentially run/update even when the dynamic elements are not visible. This depends on how the consumer is using it, but the potential is there and it could have a (negligible) impact on performance and memory. I'm not too concerned about this since it could be fixed with guidance and examples to help promote correct usage, and ensure developers are aware of it in general.

**Comment by DRiFTy17** [comment]
An initial PoC implementation of converting the existing dialog (without breaking changes) can be found here:
https://github.com/tyler-technologies-oss/forge/commit/13c691fe22d98cda44e78951ebc3331f255c9781

**Comment by DRiFTy17** [comment]
A major problem/concern with static implementations like this is related to how `position: fixed` and containing blocks work. To keep the location in the DOM, but render relative to the viewport means we should to use `position: fixed`. The problem with this is that a new containing block can easily be created on non-viewport ancestors by various properties such as `transform` or `will-change: transform` as noted [here](https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block#identifying_the_containing_block).

This is basically a deal breaker because we can't reliably position child elements relative to the viewport without developers being aware of the caveats.

Our main goal with this static implementation is related to how we want to enable an easier usage of components to avoid dynamically generating DOM on the developers behalf and appending it to the `<body>`. If we implement this static usage, developers could very easily unknowingly break layouts that cause clipping of these dynamic elements. The only way to guarantee proper positioning is to append to the `<body>`, but this comes with the current drawback of generating DOM on the fly, and more configuration properties to allow for updating content dynamically.

We really need the browsers to implement a `position` value that we can use to always force positioning relative to the initial containing block (document)...

**Reply by DRiFTy17** [reply]
We will eventually get access to adding our own elements to the [`top-layer`](https://developer.chrome.com/blog/what-is-the-top-layer/) in the browser. Native `<dialog>` elements can add to the top-layer already via the `showModal()` method, but access to adding arbitrary elements to this top-layer is not yet available. It will be the solution we need once available and implemented across browsers. For now, we should leave our implementation as-is with appending to `<body>` and revisit this at a later time.

**Comment by DRiFTy17** [comment]
The new `popover` API has been merged in the spec. It looks like we may get access to this sooner than originally planned. We can move forward with these changes now, and when browser support catches up we can make use of the `popover` attribute to force to the top-layer.

Chrome status:
https://chromestatus.com/feature/5463833265045504
https://developer.chrome.com/docs/web-platform/popover-api/

There is also a popover polyfill available already:
https://github.com/oddbird/popover-polyfill

**Reply by MikeMatusz** [reply]
I believe the tracking for Firefox is here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1808823

**Reply by DRiFTy17** [reply]
WebKit:
https://bugs.webkit.org/show_bug.cgi?id=250171

**Reply by DRiFTy17** [reply]
Safari shipped `popover` in Technology Preview 166:
https://webkit.org/blog/13964/release-notes-for-safari-technology-preview-166/

**Comment by DRiFTy17** [comment]
Proof of concept work on this has begun in the [next-overlay](https://github.com/tyler-technologies-oss/forge/tree/next-overlay) branch.

Initial assessment is positive in regards to functionality and styles, but accessibility will suffer in the short term until some newer HTML specs are merged and implemented in various browsers. Currently the accessibility aspect only works in Chrome Canary at the of writing thanks to the ARIA reflection API (Safari has it implemented, but isn't hooking the elements up in the AOM yet). More information about status of these APIs can be found [here](https://github.com/whatwg/html/pull/7934).
